### PR TITLE
HIVE-27503: Support query iceberg tag

### PR DESCRIPTION
**Pull request by zhangbutao**
### What changes were proposed in this pull request?

Support query iceberg tag like this: 

`select * from db.tbl.tag_tagName;`

In addition, Iceberg tag can not be written data and we should throw exception in compile stage if users want to write data to iceberg tag. 

### Does this PR introduce _any_ user-facing change?

No
### Is the change a dependency upgrade?

No

### How was this patch tested?

`mvn test -Dtest.output.overwrite -Dtest=TestIcebergCliDriver -Dqfile=query_iceberg_tag.q -pl itests/qtest-iceberg -Piceberg -Pitests -Drat.skip`

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 1]
```diff
@@ -2399,12 +2399,16 @@ private void getMetaData(QB qb, ReadEntity parentInput)
...
-            ts.tableHandle.getTableName(), ts.tableHandle.getBranchName());
+            ts.tableHandle.getTableName(), ts.tableHandle.getSnapshotRef());
         isTableWrittenTo |= (qb.getParseInfo().getInsertOverwriteTables().
             get(getUnescapedName((ASTNode) ast.getChild(0), ts.tableHandle.getDbName()).toLowerCase()) != null);
         assert isTableWrittenTo :
             "Inconsistent data structure detected: we are writing to " + ts.tableHandle  + " in " +
                 name + " but it's not in isInsertIntoTable() or getInsertOverwriteTables()";
+        Boolean isTableTag = Optional.ofNullable(ts.tableHandle.getSnapshotRef()).map(HiveUtils::checkTableTag)
+            .orElse(false);
+        assert isTableWrittenTo && !isTableTag : "Don't support write (insert/delete/update/merge) to iceberg tag " +
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:2410]: assert would be removed in runtime, show we throw UnsupportedOperatiionEException here?

File: storage-api/src/java/org/apache/hadoop/hive/common/TableName.java [hunk 2]
```diff
@@ -91,7 +92,7 @@ public static TableName fromString(final String name, final String defaultCatalo
       if (names.length == 2) {
         return new TableName(defaultCatalog, names[0], names[1], null);
       } else if (names.length == 3) {
-        if (names[2].startsWith(BRANCH_NAME_PREFIX)) {
+        if (names[2].startsWith(BRANCH_NAME_PREFIX) || names[2].startsWith(TAG_NAME_PREFIX)) {
```
Comment (deniskuzZ) [inline storage-api/src/java/org/apache/hadoop/hive/common/TableName.java:95]: could we use SNAPSHOT_REF matcher?

**Review by deniskuzZ — COMMENTED** [review]
+1 LGTM, minor comments

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveUtils.java [hunk 3]
```diff
@@ -443,11 +444,16 @@ public static Path getDumpPath(Path root, String dbName, String tableName) {
...
-      return branch.group(1);
+  public static String getTableSnapshotRef(String refName) {
+    Matcher ref = SNAPSHOT_REF.matcher(refName);
+    if (ref.matches()) {
+      return ref.group(1);
     }
     return null;
   }
+
+  public static Boolean checkTableTag(String refName) {
```
[the remark this answers is not in the document]
Comment (zhangbutao) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveUtils.java]: Fixed.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 1 above)
Comment (zhangbutao) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Fixed.

File: storage-api/src/java/org/apache/hadoop/hive/common/TableName.java (hunk 2 above)
Comment (zhangbutao) [reply storage-api/src/java/org/apache/hadoop/hive/common/TableName.java]: Fixed.

**Review by deniskuzZ — APPROVED** [review]
+1

**Pull request merged**